### PR TITLE
feat: edit set configmap

### PR DESCRIPTION
**Pull request by stormqueen1990**
First pass at implementing the `edit set configmap` command. Initally adding just two flags: `--from-literal` and `--new-namespace`.

Related issue: #4493

### CODE REVIEW

File: kustomize/commands/edit/add/addconfigmap.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ import (
 	"sigs.k8s.io/kustomize/api/resource"
 	"sigs.k8s.io/kustomize/api/types"
 	"sigs.k8s.io/kustomize/kustomize/v5/commands/internal/kustfile"
+	"sigs.k8s.io/kustomize/kustomize/v5/commands/internal/util"
 	"sigs.k8s.io/kustomize/kyaml/filesys"
 )
```
Comment (natasha41575) [inline kustomize/commands/edit/add/addconfigmap.go:16]: nit: Saw the  name of the file here changed, is that needed? It's already under the `add` subdirectory, so not sure that we need to re-specify `add` in the file name

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go [hunk 2]
```diff
@@ -1,54 +1,57 @@
 // Copyright 2019 The Kubernetes Authors.
 // SPDX-License-Identifier: Apache-2.0
 
-package add
+package util
```
Comment (natasha41575) [inline kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go:4]: suggestion: separate out this change in its own PR, so that this PR focuses on just adding the `edit set configmap` feature

**Review by natasha41575 — COMMENTED** [review]
Quick drive by comments. Overall approach LGTM. I can do a more thorough review once this is taken out of draft state.

File: kustomize/commands/edit/add/addconfigmap.go (hunk 1 above)
Comment (stormqueen1990) [reply kustomize/commands/edit/add/addconfigmap.go:16]: Not needed, but I think it'd be nice to have for consistency with the other files in the same folder.

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go (hunk 2 above)
Comment (stormqueen1990) [reply kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go:4]: Will do!

Comment (stormqueen1990) [reply kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go:4]: Extracted this to a new PR: #5409. Also undid the renames.

File: kustomize/commands/edit/add/addconfigmap.go (hunk 1 above)
Comment (stormqueen1990) [reply kustomize/commands/edit/add/addconfigmap.go:16]: I undid this in the new #5409 PR.

Comment (natasha41575) [reply kustomize/commands/edit/add/addconfigmap.go:16]: Oh I see, that's a good point - didn't realize that the other files are doing it this way as well. In that case I agree with you that it is good to have the consistency; feel free to add the change back if you'd like.

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go (hunk 2 above)
Comment (natasha41575) [reply kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go:4]: Approved. Thanks!

Comment (stormqueen1990) [reply kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go]: Rebased this to pick up the new changes!

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs_test.go [hunk 3]
```diff
@@ -14,13 +14,13 @@ import (
 
 func TestDataValidation_NoName(t *testing.T) {
 	fa := ConfigMapSecretFlagsAndArgs{}
-	require.Error(t, fa.Validate([]string{}))
+	require.Error(t, fa.ValidateAdd([]string{}))
```
Comment (ncapps) [inline kustomize/commands/internal/util/configmapSecretFlagsAndArgs_test.go:17]: It would be nice if tests for `ValidateSet` and `UpdateLiteralSources` were added.

File: kustomize/commands/edit/set/configmap.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+)
+
+func NewCmdSetConfigMap(
+	fSys filesys.FileSystem,
+	ldr ifc.KvLoader,
+	rf *resource.Factory,
+) *cobra.Command {
+	var flags util.ConfigMapSecretFlagsAndArgs
+	cmd := &cobra.Command{
+		Use:   "configmap NAME [--from-literal=key1=value1]",
```
Comment (ncapps) [inline kustomize/commands/edit/set/configmap.go:26]: Should `Use` specify the optional `--namespace` and `--new-namespace` flags?

File: kustomize/commands/edit/set/configmap.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
+// Copyright 2023 The Kubernetes Authors.
```
Comment (ncapps) [inline kustomize/commands/edit/set/configmap.go:1]: All other filenames in the `kustomize/commands/edit/set` directory begin with `set`. Would it be more consistent to name this file `setconfigmap.go`?

**Review by ncapps — COMMENTED** [review]
I left minor questions and suggestions. I assume the `set secret` functionality will be added in a follow up PR?

File: kustomize/commands/edit/set/configmap.go (hunk 5 above)
Comment (stormqueen1990) [reply kustomize/commands/edit/set/configmap.go:1]: Yes, that's currently following the same standard as the `add` folder. I plan to open a subsequent PR to change these names so they are all standardized.

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs_test.go (hunk 3 above)
Comment (stormqueen1990) [reply kustomize/commands/internal/util/configmapSecretFlagsAndArgs_test.go]: Thanks for flagging this!! I totally forgot 😅 added some tests, could you please take a look and check if they address your concerns?

File: kustomize/commands/edit/set/configmap.go (hunk 4 above)
Comment (stormqueen1990) [reply kustomize/commands/edit/set/configmap.go]: Absolutely yes, just updated this to include these flags as optionals.

**Comment by stormqueen1990** [comment]
> I left minor questions and suggestions. I assume the `set secret` functionality will be added in a follow up PR?

Hi @ncapps! Yes, that's correct. I intend to add `edit set secret` in a separate PR, as smaller blocks are usually easier to review and push forward.

**Review by charles-chenzz — COMMENTED** [review]
overall lgtm
/lgtm

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs_test.go (hunk 3 above)
Comment (ncapps) [reply kustomize/commands/internal/util/configmapSecretFlagsAndArgs_test.go]: Thanks!

**Comment by ncapps** [comment]
I am not an official reviewer but these changes lgtm.

File: kustomize/commands/edit/set/configmap.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+// findConfigMapArgs finds the generator arguments corresponding to the specified
+// ConfigMap name. ConfigMap must exist for this command to be successful.
+func findConfigMapArgs(m *types.Kustomization, name, namespace string) (*types.ConfigMapArgs, error) {
+	cmIndex := slices.IndexFunc(m.ConfigMapGenerator, func(cmArgs types.ConfigMapArgs) bool {
+		return name == cmArgs.Name && namespace == cmArgs.Namespace
+	})
+
+	if cmIndex == -1 {
+		return nil, fmt.Errorf("unable to find ConfigMap with name '%s'", name)
```
[the remark this answers is not in the document]
Comment (varshaprasad96) [reply kustomize/commands/edit/set/configmap.go:147]: Just curious, can cmArgs ever be nil? Asking this because `types.Kustomization` is being passed as a pointer. Would it be safer to check if it is nil first and return it before itself?

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go [hunk 7]
```diff
@@ -62,32 +68,49 @@ func (a *ConfigMapSecretFlagsAndArgs) Validate(args []string) error {
...
+// - some/dir/ant
+// - some/dir/apple
+// - bfile=some/dir/banana
 // i.e. everything is converted to a key=value pair,
 // where the value is always a relative file path,
 // and the key, if missing, is the same as the value.
 // In the case where the key is explicitly declared,
 // the globbing, if present, must have exactly one match.
 func (a *ConfigMapSecretFlagsAndArgs) ExpandFileSource(fSys filesys.FileSystem) error {
+	const NumberComponentsWithKey = 2
```
Comment (varshaprasad96) [inline kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go:104]: Nit:
```suggestion
	const numberComponentsWithKey = 2
```
Shouldn't matter since this is a local var, but we need not export it.

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go [hunk 8]
```diff
@@ -114,6 +137,49 @@ func (a *ConfigMapSecretFlagsAndArgs) ExpandFileSource(fSys filesys.FileSystem)
...
+// UpdateLiteralSources looks for literal sources that already exist and tries
+// to replace their values with new values.
+// The key specified must exist in the target resource (ConfigMap or Secret).
+func UpdateLiteralSources(
+	args *types.GeneratorArgs,
+	flags ConfigMapSecretFlagsAndArgs,
+) error {
+	sources := make(map[string]any)
+	for _, val := range args.LiteralSources {
+		key, value, found := strings.Cut(val, "=")
```
Comment (varshaprasad96) [inline kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go:149]: Not necessary, but trying to be super careful since we are dealing with strings here, and users can manipulate the input in any way. Do we want to first check the occurrences of "=" to verify it is appearing once, and return early if it is a malformed string?

File: kustomize/commands/edit/set/configmap.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (stormqueen1990) [reply kustomize/commands/edit/set/configmap.go:147]: @varshaprasad96 I'd say it's unlikely, because `ConfigMapGenerator` itself is not a pointer (it's a slice within the kustomization file). I'd expect the read for `kustomization.yaml` to fail before we can have a `nil` kustomization object, but to be honest I can't picture a situation where the evaluated value of the object in `kustomization.yaml` would result in a `nil` slice off the top of my head.

Do you have a specific use case and/or failure mode in mind? I'm happy to add some tests to safeguard against this.

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go (hunk 7 above)
Comment (stormqueen1990) [reply kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go:104]: I intended this to be a constant, since it used to be a magic number before (and the linter was complaining 😅). I agree with not exporting, but changing this to a var can be problematic if any future changes accidentally modify this value.

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go (hunk 8 above)
Comment (stormqueen1990) [reply kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go:149]: Yes, I think that'd be a good idea.

Comment (stormqueen1990) [reply kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go]: Updated!

File: kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go (hunk 7 above)
Comment (stormqueen1990) [reply kustomize/commands/internal/util/configmapSecretFlagsAndArgs.go]: Updated!

**Review by charles-chenzz — COMMENTED** [review]
/lgtm

File: kustomize/commands/edit/set/configmap.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (varshaprasad96) [reply kustomize/commands/edit/set/configmap.go]: Got it! Thanks @stormqueen1990! I was just curious because of the func was using a pointer, but makes sense that as we already have a fail safe knob before :)

File: kustomize/commands/edit/add/configmap.go [hunk 9]
```diff
@@ -142,6 +143,10 @@ func addConfigMap(
 	// Validate by trying to create corev1.configmap.
 	args.Options = types.MergeGlobalOptionsIntoLocal(
 		args.Options, k.GeneratorOptions)
+
+	// guarantee end order is predictable
+	slices.Sort(args.LiteralSources)
```
Comment (natasha41575) [inline kustomize/commands/edit/add/configmap.go:148]: Do you mind explaining the motivation here? It seems to me that preserving the order of the user's original input might be more desirable than sorting it

File: kustomize/commands/edit/set/configmap.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+	}
+
+	return nil
+}
+
+// findConfigMapArgs finds the generator arguments corresponding to the specified
+// ConfigMap name. ConfigMap must exist for this command to be successful.
+func findConfigMapArgs(m *types.Kustomization, name, namespace string) (*types.ConfigMapArgs, error) {
+	cmIndex := slices.IndexFunc(m.ConfigMapGenerator, func(cmArgs types.ConfigMapArgs) bool {
+		return name == cmArgs.Name && namespace == cmArgs.Namespace
```
Comment (natasha41575) [inline kustomize/commands/edit/set/configmap.go:143]: nit: This is an edge case, but I think we should try to handle the default namespace being specified as either as an empty string or default. 

`""` and `"default"` should be treated as the same namespace, so, for example, it should still match if there is no namespace specified in `cmArgs.Namespace` but `flags.Namespace` is `default` and vice versa.

We should also have a test or two for this.

**Review by natasha41575 — COMMENTED** [review]
Overall looks good. Minor question and nit.

File: kustomize/commands/edit/set/configmap.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (stormqueen1990) [reply kustomize/commands/edit/set/configmap.go]: No problem @varshaprasad96! Do you feel this is resolved? Gauging to see if I can click "resolve conversation" 😃

File: kustomize/commands/edit/add/configmap.go (hunk 9 above)
Comment (stormqueen1990) [reply kustomize/commands/edit/add/configmap.go:148]: This is mostly for testing purposes 😅 I can roll it back.

File: kustomize/commands/edit/set/configmap.go (hunk 10 above)
Comment (stormqueen1990) [reply kustomize/commands/edit/set/configmap.go:143]: That's fair. Will update.

File: kustomize/commands/edit/set/configmap.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (varshaprasad96) [reply kustomize/commands/edit/set/configmap.go]: Yes, resolved from my end :)

File: kustomize/commands/edit/set/configmap.go (hunk 10 above)
Comment (stormqueen1990) [reply kustomize/commands/edit/set/configmap.go:143]: I just found out that this is already an issue for `edit add configmap`/`edit add secret`. Will open a separate PR to fix the behaviour on these two other places.

Comment (stormqueen1990) [reply kustomize/commands/edit/set/configmap.go:143]: I opened #5430 to address this in other places and will hold this one until the other gets merged, so I can reuse the same method.

/hold

File: kustomize/commands/edit/add/configmap.go (hunk 9 above)
Comment (stormqueen1990) [reply kustomize/commands/edit/add/configmap.go]: This was rolled back.

File: kustomize/commands/edit/set/configmap.go (hunk 10 above)
Comment (stormqueen1990) [reply kustomize/commands/edit/set/configmap.go]: #5430 was merged, so I applied the changes to this code to handle the empty and the default namespaces as being the same.

**Review by charles-chenzz — COMMENTED** [review]
Thanks
/lgtm
